### PR TITLE
MueLu: clean-up MueLu_AlgebraicPermutationStrategy_def.hpp for #3492

### LINKED ISSUE #3492: MueLu: AlgebraicPermutationStrategy does reductions on scalar instead of a magnitude type
@trilinos/muelu 
@mhoemmen 

## Expectations
MueLu_AlgebraicPermutationStrategy_def.hpp should not perform reductions on `Scalar`, especially with the `REDUCE_MIN/MAX` closure. Instead perform said reductions on `Teuchos::ScalarTraits<Scalar>::magnitudeType`. This will alleviate issues of performing a reduction on `std::complex<>`.

## Current Behavior
Currently it is unclear that the results coming out of the MPI reductions are probably not useful when `Scalar=std::complex<>`.

## Motivation and Context
This is a bug it needs to be fixed...

## Possible Solution
Carefully adjust types in the class so that all reductions are performed on `magnitudeType` if possible.

## Related Issues
* Follows #3480
* Precedes 
* Related to 
* Part of 
* Composed of

### PR DESCRIPTION
**Pull request by lucbv**
@trilinos/muelu 

## Description
Changed the type of some variables from `Scalar` to `Teuchos::ScalarTraits<Scalar>::magnitudeType` to avoid issues related reductions using: REDUCE_MIN/MAX operations that are not well defined when `Scalar=std::complex<>`.

## Motivation and Context
This cleans-up the code and potentially fixes a bug

## Related Issues

* Closes #3492

## How Has This Been Tested?
A local build was done with `Scalar=std::complex<>` has been performed

## Checklist

- [x] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.

### CODE REVIEW

**Comment by jhux2** [comment]
*[a discussion](https://reviewable.io/reviews/trilinos/trilinos/4369#-LYSfqiCExti4A86Sysv:-LYSfqiD4D8O_wW-pMj2:b-s2apzq) (no related file):*
This looks good to me, hopefully didn't miss anything due to whitespace changes.

---

File: packages/muelu/src/Utils/MueLu_AlgebraicPermutationStrategy_def.hpp [hunk 1]
```diff
@@ -31,674 +31,690 @@
...
+      //
+    MueLu_sumAll(gDomVec->getMap()->getComm(), (LocalOrdinal)localMultColRequests, globalMultColRequests);
+
+    if(globalMultColRequests > 0) {
+      // special handling: two processors request the same global column id.
+      // decide which processor gets it
+
+      // distribute number of multipleColRequests to all processors
+      // each processor stores how many column ids for exchange are handled by the cur proc
+      std::vector<GlobalOrdinal> numMyMultColRequests(numProcs,0);
```
Comment (mhoemmen) [inline packages/muelu/src/Utils/MueLu_AlgebraicPermutationStrategy_def.hpp]: There must be a better way to communicate this information than an all-reduce on a whole array, but that's a separate issue.  (Maybe Distributor already encapsulates this communication pattern for you?)

File: packages/muelu/src/Utils/MueLu_AlgebraicPermutationStrategy_def.hpp [hunk 2]
```diff
@@ -31,674 +31,690 @@
...
+      // loop over all local column GIDs that are also requested by other procs
+      for(size_t i = 0; i < multipleColRequests.size(); i++) {
+        procMultRequestedColIds[nMyOffset + i] = multipleColRequests[i]; // all weights are > 0 ?
+      }
 
-    // loop over global_procOverlappingWeights and eliminate wrong entries...
-    for (size_t k = 0; k<global_procMultRequestedColIds.size(); k++) {
-      GlobalOrdinal globColId = global_procMultRequestedColIds[k];
+      // template ordinal, package (double)
+      Teuchos::reduceAll(*comm, Teuchos::REDUCE_MAX, Teuchos::as<int>(globalMultColRequests),
```
Comment (mhoemmen) [inline packages/muelu/src/Utils/MueLu_AlgebraicPermutationStrategy_def.hpp]: I'm kinda scared by this line of code, but it's not in scope of this PR.

File: packages/muelu/src/Utils/MueLu_AlgebraicPermutationStrategy_def.hpp [hunk 3]
```diff
@@ -31,674 +31,690 @@
...
-            if((*p).second == globColId)
-              p = permutedDiagCandidatesFiltered.erase(p);
-            else
-              p++;
+        Teuchos::reduceAll(*comm, Teuchos::REDUCE_MAX, numProcs,
+                           &MyWeightForColId[0],
+                           &GlobalWeightForColId[0]);
+
+        if(gColVec->getMap()->isNodeGlobalElement(globColId)) {
+          // note: 2 procs could have the same weight for a column index.
```
Comment (mhoemmen) [inline packages/muelu/src/Utils/MueLu_AlgebraicPermutationStrategy_def.hpp]: Again, this is not in scope of this PR, but Distributor and Directory might do a lot of this work for you in a more communication-efficient way.

**Pull request merged**